File: 2022/day2a_rock_paper_scissors.swift

enum Choice: Equatable {
    case rock
    case paper
    case scissors
}

func score(_ choice: Choice) -> Int {
    switch choice {
    case .rock:     return 1
    case .paper:    return 2
    case .scissors: return 3
    }
}

func getChoice(_ s: Substring) -> Choice {
    switch s {
    case "A", "X": return .rock
    case "B", "Y": return .paper
    case "C", "Z": return .scissors
    default: unreachable()
    }
}

func unreachable() -> Never {
    fatalError("unreachable")
}

var total = 0
while let line = readLine() {
    let parts = line.split(separator: " ")
    let opponent = parts[0]
    let your = parts[1]

    let opponent_choice = getChoice(opponent)
    let your_choice = getChoice(your)

    switch (opponent_choice, your_choice) {
    case (.rock, .paper),
         (.paper, .scissors),
         (.scissors, .rock):
        total += 6
    case let (a, b) where a == b:
        total += 3
    default: break
    }

    total += score(your_choice)
}

print("\(total)")

File: 2022/day2a_rock_paper_scissors.hs
import Data.Char (isSpace)
import Control.Arrow ((>>>))
import Data.Function ((&))

data Choice = Rock | Paper | Scissors deriving Eq

instance Read Choice where
    readsPrec _ s = do choice <- case stripped of
                                   'A':_ -> pure Rock
                                   'X':_ -> pure Rock
                                   'B':_ -> pure Paper
                                   'Y':_ -> pure Paper
                                   'C':_ -> pure Scissors
                                   'Z':_ -> pure Scissors
                                   []    -> []
                       return (choice, tail stripped)
        where stripped = s & span isSpace & snd

Rock     `beats` Scissors = True
Paper    `beats` Rock     = True
Scissors `beats` Paper    = True
_        `beats` _        = False

choiceScore Rock     = 1
choiceScore Paper    = 2
choiceScore Scissors = 3

score opponentsChoice yourChoice = choiceScore yourChoice + matchResultScore
    where matchResultScore = if yourChoice `beats` opponentsChoice
                               then 6
                               else if yourChoice == opponentsChoice
                               then 3
                               else 0

main = interact
     ( lines
   >>> fmap ( words
          >>> fmap read
          >>> \[opponent, you] -> score opponent you
            )
   >>> sum
   >>> show
   >>> (++"\n")
     )
